libfreerdp/codec/rfx_neon.c

```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
+				tmp_n = vaddq_s16(tmp_n, h_n_m);
+			}
+
+			tmp_n = vshrq_n_s16(tmp_n, 1);
+			int16x8_t dst_n = vsubq_s16(l_n, tmp_n);
+			vst1q_s16(dst_ptr, dst_n);
+			l_ptr += 8;
+			h_ptr += 8;
+			dst_ptr += 8;
+		}
+
+		if (nDstCount > batchSize)
+		{
+			int16_t h_n = (n == 31) ? *(h_ptr - nHighStep) : *h_ptr;
+			int16_t tmp_n = h_n + 1;
+			if (n == 0)
+			{
+				tmp_n += h_n;
+			}
+			else if (n < 31)
+			{
+				tmp_n += *(h_ptr - nHighStep);
+			}
+			tmp_n >>= 1;
+			*dst_ptr = *l_ptr - tmp_n;
+			l_ptr += 1;
+			h_ptr += 1;
+			dst_ptr += 1;
+		}
+
+		dst_ptr += nDstStep;
+	}
+
+	if (forceBandSize < 32)
+	{
+		for (x = 0; x < batchSize; x += 8)
+		{
+			int16x8_t l_n = vld1q_s16(l_ptr);
+			int16x8_t h_n = vld1q_s16(h_ptr - nHighStep);
+			int16x8_t tmp_n = vsubq_s16(l_n, h_n);
+			vst1q_s16(dst_ptr, tmp_n);
+			l_ptr += 8;
+			h_ptr += 8;
+			dst_ptr += 8;
+		}
+
+		if (nDstCount > batchSize)
+		{
+			*dst_ptr = *l_ptr - *(h_ptr - nHighStep);
+			l_ptr += 1;
+			h_ptr += 1;
+			dst_ptr += 1;
+		}
+	}
+
+	h_ptr = pHighBand;
+	dst_ptr = pDstBand + nDstStep;
+
+	/* Odd coefficients */
+	for (n = 0; n < forceBandSize; n++)
+	{
+		for (x = 0; x < batchSize; x += 8)
+		{
+			// dst[2n + 1] = (h[n] << 1) + ((dst[2n] + dst[2n + 2]) >> 1);
+			int16x8_t tmp_n = vld1q_s16(dst_ptr - nDstStep);
+			if (n == 31)
+			{
+				int16x8_t dst_n_p = vld1q_s16(l_ptr);
+				l_ptr += 8;
+				tmp_n = vaddq_s16(tmp_n, dst_n_p);
+				tmp_n = vshrq_n_s16(tmp_n, 1);
+			}
+			else
+			{
+				int16x8_t dst_n_p = vld1q_s16(dst_ptr + nDstStep);
+				tmp_n = vaddq_s16(tmp_n, dst_n_p);
+				tmp_n = vshrq_n_s16(tmp_n, 1);
+				int16x8_t h_n = vld1q_s16(h_ptr);
+				h_n = vshlq_n_s16(h_n, 1);
+				tmp_n = vaddq_s16(tmp_n, h_n);
+			}
+			vst1q_s16(dst_ptr, tmp_n);
+			h_ptr += 8;
+			dst_ptr += 8;
+		}
+
+		if (nDstCount > batchSize)
+		{
+			int16_t tmp_n = *(dst_ptr - nDstStep);
+			if (n == 31)
+			{
+				int16_t dst_n_p = *l_ptr;
+				l_ptr += 1;
+				tmp_n += dst_n_p;
+				tmp_n >>= 1;
+			}
+			else
+			{
+				int16_t dst_n_p = *(dst_ptr + nDstStep);
+				tmp_n += dst_n_p;
+				tmp_n >>= 1;
+				int16_t h_n = *h_ptr;
+				h_n <<= 1;
+				tmp_n += h_n;
+			}
+			*dst_ptr = tmp_n;
+			h_ptr += 1;
+			dst_ptr += 1;
+		}
+
+		dst_ptr += nDstStep;
+	}
+}
+
+static INLINE size_t prfx_get_band_l_count(size_t level)
+{
+	return (64 >> level) + 1;
+}
+
+static INLINE size_t prfx_get_band_h_count(size_t level)
+{
+	if (level == 1)
+		return (64 >> 1) - 1;
+	else
+		return (64 + (1 << (level - 1))) >> level;
+}
+
+static INLINE void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
+rfx_dwt_2d_decode_extrapolate_block_neon(INT16* buffer, INT16* temp, size_t level)
+{
+	size_t nDstStepX;
+	size_t nDstStepY;
+	INT16 *HL, *LH;
+	INT16 *HH, *LL;
+	INT16 *L, *H, *LLx;
+
+	const size_t nBandL = prfx_get_band_l_count(level);
+	const size_t nBandH = prfx_get_band_h_count(level);
+	size_t offset = 0;
+
+	HL = &buffer[offset];
```

Comment:
you can declare the variables mid code now (`C99`) and better make them `const` if you do not intend to modify them